### PR TITLE
feat: Add signed and unprotected JWT [SIW-283]

### PR DESCRIPTION
**Pull request by grausof**
## Short description
Adds support functions to:
- Create a JWT (`producer.ts`)
- Create an unprotected JWT (`unprotected.ts`)
- Create a signed JWT (`sign.ts`)

It also adds functions for base64 encoding and decoding

## How to test
Unit test!

### CODE REVIEW

File: src/__tests__/sign.test.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  kid: 'EC#1',
+  alg: 'ES256',
+};
+
+const signature =
+  '6wA0M6rNYNSFN_EylzMA6ElAibW7FVSZyoLNEkHU5c_RKuiNenT08YIMvbysYautLZotUedEMP5xCyNpY34x6Q';
+
+describe('Sign JWT', function () {
+  it('must be signed correctly', async () => {
+    let signedJwt = new SignJWT(jwtPayload)
```
Comment (hevelius) [inline src/__tests__/sign.test.ts]:
```suggestion
    const signedJwt = new SignJWT(jwtPayload)
```

File: src/__tests__/sign.test.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  typ: 'jwt',
+  kid: 'EC#1',
+  alg: 'ES256',
+};
+
+const signature =
+  '6wA0M6rNYNSFN_EylzMA6ElAibW7FVSZyoLNEkHU5c_RKuiNenT08YIMvbysYautLZotUedEMP5xCyNpY34x6Q';
+
+describe('Sign JWT', function () {
+  it('must be signed correctly', async () => {
```
Comment (hevelius) [inline src/__tests__/sign.test.ts]:
```suggestion
  it('it should be signed correctly', async () => {
```

File: src/__tests__/sign.test.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    let signedJwt = new SignJWT(jwtPayload)
+      .setProtectedHeader(jwtHeader)
+      .sign(signature);
+    expect(signedJwt).toBe(
+      'eyJ0eXAiOiJqd3QiLCJraWQiOiJFQyMxIiwiYWxnIjoiRVMyNTYifQ.eyJpc3MiOiJkZW1vIiwic3ViIjoiZGVtbyIsImlhdCI6MTY3NTIwNjAwMCwiZXhwIjoxNzA2NzQyMDAwfQ.6wA0M6rNYNSFN_EylzMA6ElAibW7FVSZyoLNEkHU5c_RKuiNenT08YIMvbysYautLZotUedEMP5xCyNpY34x6Q'
+    );
+  });
+  it('signature must fail', async () => {
```
Comment (hevelius) [inline src/__tests__/sign.test.ts]:
```suggestion
  it('it should fails with empty string', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  sub: 'demo',
+  iat: 1675206000,
+  exp: 1706742000,
+};
+
+const encodedUnsecureJwt =
+  'eyJhbGciOiJub25lIn0.eyJpc3MiOiJkZW1vIiwic3ViIjoiZGVtbyIsImlhdCI6MTY3NTIwNjAwMCwiZXhwIjoxNzA2NzQyMDAwfQ.';
+
+describe('Unsecured JWT', function () {
+  it('must be encoded correctly', async () => {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
  it('it should be encoded correctly', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    let unsecured = new UnsecuredJWT(jwtPayload).encode();
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
    const unsecured = new UnsecuredJWT(jwtPayload).encode();
```

File: src/__tests__/unsecured.test.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    expect(unsecured).toBe(encodedUnsecureJwt);
+  });
+  it('must fail the encode', async () => {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
  it('it should fail encoding with unknown payload', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    expect(() =>
+      new UnsecuredJWT('' as unknown as JWTPayload).encode()
+    ).toThrowError(TypeError);
+  });
+  it('must be decoded correctly', async () => {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
  it('it should be decoded correctly', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    let unsecured = UnsecuredJWT.decode(encodedUnsecureJwt, {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
    const unsecured = UnsecuredJWT.decode(encodedUnsecureJwt, {
```

File: src/__tests__/unsecured.test.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+      currentDate: new Date(2023, 7, 14, 0, 0),
+    });
+    expect(unsecured.payload).toMatchObject(jwtPayload);
+  });
+  it('must fail the decode', async () => {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
  it('it should fail the decode with an empty string', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    expect(() => UnsecuredJWT.decode('')).toThrowError(JWTInvalid);
+  });
+  it('must return content to sign', async () => {
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
  it('it should return the right content to sign', async () => {
```

File: src/__tests__/unsecured.test.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    let toSign = new UnsecuredJWT(jwtPayload).toSign();
```
Comment (hevelius) [inline src/__tests__/unsecured.test.ts]:
```suggestion
    const toSign = new UnsecuredJWT(jwtPayload).toSign();
```

File: src/utils/base64.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+import { btoa, atob } from 'abab';
+import { JOSEError } from './errors';
+
+export function encodeBase64(value: string): string {
+  let encoded = btoa(value);
+  if (encoded) {
+    // eslint-disable-next-line no-div-regex
+    return encoded.replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_');
```
Comment (hevelius) [inline src/utils/base64.ts:20]: Not clear for me why we need to use replace regex rule here.

**Review by hevelius — CHANGES_REQUESTED** [review]

**Review by balanza — COMMENTED** [review]
I wonder why you choose the OOP paradigm 🤷

File: src/utils/base64.ts (hunk 12 above)
Comment (grausof) [reply src/utils/base64.ts:20]: By default, the `replace()` method replaces the first match if the regexp doesn’t use the global flag (`g`). To replace all matches, you use the global flag (`g`) in the regexp.

Comment (hevelius) [reply src/utils/base64.ts:20]: Not clear to me is why we don't use encoded directly instead refer replace.

Comment (grausof) [reply src/utils/base64.ts:20]: In Base64 encoding, the length of an output-encoded String must be a multiple of four. If necessary, the encoder adds one or two padding characters `(=)` at the end of the output as needed in order to meet this requirement. 

JWTs make use of the base64url encoding as defined in RFC 4648 [RFC4648] . As allowed by Section 3.2 of the RFC, this specification mandates that base64url encoding when used with JWTs MUST NOT use padding. The reason for this restriction is that the padding character ('=') is not URL safe. 

Maybe I should add a comment!

Comment (hevelius) [reply src/utils/base64.ts:20]: Yep it should be a good idea move your reply in a comment.

Comment (grausof) [reply src/utils/base64.ts:20]: done!

**Comment by grausof** [comment]
> I wonder why you choose the OOP paradigm 🤷

OOP was only used for creating the JWTs. This for two reasons:
1. To make jose compatible interfaces for node
2. Because this allows me to create an object quickly with the following syntax:

```
let jwtToSign = new UnsecuredJWT({ metadata: 'demo' })
  .setAudience('demo')
  .setExpirationTime('1h')
  .setIssuedAt()
  .toSign();
```

**Review by hevelius — APPROVED** [review]
LGTM

**Pull request merged**